lib/rubocop/cop/rails/redundant_presence_validation_on_belongs_to.rb

```diff
@@ -0,0 +1,192 @@
+      #   belongs_to :user
+      #   validates :user_id, presence: true
+      #
+      #   # bad
+      #   belongs_to :author, foreign_key: :user_id
+      #   validates :user_id, presence: true
+      #
+      #   # good
+      #   belongs_to :user
+      #
+      #   # good
+      #   belongs_to :author, foreign_key: :user_id
+      #
+      class RedundantPresenceValidationOnBelongsTo < Base
+        include RangeHelp
+        extend AutoCorrector
+        extend TargetRailsVersion
+
+        MSG = 'Remove explicit presence validation for `%<association>s`.'
+        RESTRICT_ON_SEND = %i[validates].freeze
+
+        minimum_target_rails_version 5.0
+
+        # @!method presence_validation?(node)
+        #   Match a `validates` statement with a presence check
+        #
+        #   @example source that matches - by association
+        #     validates :user, presence: true
+        #
+        #   @example source that matches - with presence options
+        #     validates :user, presence: { message: 'duplicate' }
+        #
+        #   @example source that matches - by a foreign key
+        #     validates :user_id, presence: true
+        def_node_matcher :presence_validation?, <<~PATTERN
+          $(
+            send nil? :validates
+            (sym $_)
+            ...
+            $(hash <$(pair (sym :presence) {true hash}) ...>)
+          )
+        PATTERN
+
+        # @!method optional_option?(node)
+        #  Match a `belongs_to` association with an optional option in a hash
+        def_node_matcher :optional?, <<~PATTERN
+          (send nil? :belongs_to _ ... #optional_option?)
+        PATTERN
+
+        # @!method optional_option?(node)
+        #  Match an optional option in a hash
+        def_node_matcher :optional_option?, <<~PATTERN
+          {
+            (hash <(pair (sym :optional) true) ...>)   # optional: true
+            (hash <(pair (sym :required) false) ...>)  # required: false
+          }
+        PATTERN
+
+        # @!method any_belongs_to?(node, association:)
+        #   Match a class with `belongs_to` with no regard to `foreign_key` option
+        #
+        #   @example source that matches
+        #     belongs_to :user
+        #
+        #   @example source that matches - regardless of `foreign_key`
+        #     belongs_to :author, foreign_key: :user_id
+        #
+        #   @param node [RuboCop::AST::Node]
+        #   @param association [Symbol]
+        #   @return [Array<RuboCop::AST::Node>, nil] matching node
+        def_node_matcher :any_belongs_to?, <<~PATTERN
+          (begin
+            <
+              $(send nil? :belongs_to (sym %association) ...)
+              ...
+            >
+          )
+        PATTERN
+
+        # @!method belongs_to?(node, key:, fk:)
+        #   Match a class with a matching association, either by name or an explicit
+        #   `foreign_key` option
+        #
+        #   @example source that matches - fk matches `foreign_key` option
```

Comment:
Can you align the indents of `@example`, `@param`, and `@return` to `@!method`?

Reply:
I don't quite understand the note, but it appears that it's fine as is according to [YARD doc](https://rubydoc.info/gems/yard/file/docs/Tags.md#method). What do you think?

>  Note:

> For backwards compatibility support, you do not need to indent the method's docstring text. If a `@!method` directive is seen with no indented block, the entire docstring is used as the new method's docstring text.

Reply:
Ah, I get it. There is no problem. Thank you for the documentation link.